File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let temp = new ListNode();
    let current = temp;
    let carry = 0;

    while (l1 || l2 || carry) {
        const calculationResult = ((l1?.val || 0) + (l2?.val || 0) + carry)
        current.next = new ListNode(calculationResult % 10);
        carry = ~~(calculationResult/10)
        l1 = l1?.next
        l2 = l2?.next
        current = current.next;
    }

    return temp.next;


};

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        head = ListNode()
        current = head
        carry = 0

        while True:
            if not l1 and not l2:
                break
            remaining = l1 or l2
            s = carry + (l1.val if l1 else 0) + (l2.val if l2 else 0)
            remaining.val = s % 10
            carry = s // 10
            current.next = remaining
            if l1: l1 = l1.next
            if l2: l2 = l2.next
            current = current.next
        if carry > 0:
            current.next = ListNode(carry)
        return head.next
            
